tests/sha1_test.c

```diff
@@ -0,0 +1,256 @@
+    struct aws_byte_cursor expected_buf = aws_byte_cursor_from_array(expected, sizeof(expected));
+
+    return s_verify_hash_test_case(allocator, &input, &expected_buf, aws_sha1_new);
+}
+
+AWS_TEST_CASE(sha1_nist_test_case_4, s_sha1_nist_test_case_4_fn)
+
+static int s_sha1_nist_test_case_5_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+
+    aws_cal_library_init(allocator);
+
+    struct aws_hash *hash = aws_sha1_new(allocator);
+    ASSERT_NOT_NULL(hash);
+    struct aws_byte_cursor input = aws_byte_cursor_from_c_str("a");
+
+    for (size_t i = 0; i < 1000000; ++i) {
+        ASSERT_SUCCESS(aws_hash_update(hash, &input));
+    }
+
+    uint8_t output[AWS_SHA1_LEN] = {0};
+    struct aws_byte_buf output_buf = aws_byte_buf_from_array(output, sizeof(output));
+    output_buf.len = 0;
+    ASSERT_SUCCESS(aws_hash_finalize(hash, &output_buf, 0));
+
+    uint8_t expected[] = {
+        0x34, 0xaa, 0x97, 0x3c, 0xd4, 0xc4, 0xda, 0xa4, 0xf6, 0x1e,
+        0xeb, 0x2b, 0xdb, 0xad, 0x27, 0x31, 0x65, 0x34, 0x01, 0x6f,
+    };
+    struct aws_byte_cursor expected_buf = aws_byte_cursor_from_array(expected, sizeof(expected));
+    ASSERT_BIN_ARRAYS_EQUALS(expected_buf.ptr, expected_buf.len, output_buf.buffer, output_buf.len);
+
+    aws_hash_destroy(hash);
+
+    aws_cal_library_clean_up();
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(sha1_nist_test_case_5, s_sha1_nist_test_case_5_fn)
+
+static int s_sha1_nist_test_case_5_truncated_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+
+    aws_cal_library_init(allocator);
+
+    struct aws_hash *hash = aws_sha1_new(allocator);
+    ASSERT_NOT_NULL(hash);
+    struct aws_byte_cursor input = aws_byte_cursor_from_c_str("a");
+
+    for (size_t i = 0; i < 1000000; ++i) {
+        ASSERT_SUCCESS(aws_hash_update(hash, &input));
+    }
+
+    uint8_t expected[] = {
+        0x34, 0xaa, 0x97, 0x3c, 0xd4, 0xc4, 0xda, 0xa4, 0xf6, 0x1e, 0xeb, 0x2b, 0xdb, 0xad, 0x27, 0x31};
+    struct aws_byte_cursor expected_buf = aws_byte_cursor_from_array(expected, sizeof(expected));
+    uint8_t output[AWS_SHA1_LEN] = {0};
+    struct aws_byte_buf output_buf = aws_byte_buf_from_array(output, expected_buf.len);
+    output_buf.len = 0;
+    ASSERT_SUCCESS(aws_hash_finalize(hash, &output_buf, 16));
+
+    ASSERT_BIN_ARRAYS_EQUALS(expected_buf.ptr, expected_buf.len, output_buf.buffer, output_buf.len);
+
+    aws_hash_destroy(hash);
+
+    aws_cal_library_clean_up();
+
+    return AWS_OP_SUCCESS;
+}
+
+AWS_TEST_CASE(sha1_nist_test_case_5_truncated, s_sha1_nist_test_case_5_truncated_fn)
+
+static int s_sha1_nist_test_case_6_fn(struct aws_allocator *allocator, void *ctx) {
+    (void)ctx;
+
+    aws_cal_library_init(allocator);
+
+    struct aws_hash *hash = aws_sha1_new(allocator);
+    ASSERT_NOT_NULL(hash);
+    struct aws_byte_cursor input =
+        aws_byte_cursor_from_c_str("abcdefghbcdefghicdefghijdefghijkefghijklfghijklmghijklmnhijklmno");
+
+    for (size_t i = 0; i < 16777216; ++i) {
```

Comment:
Same comment as above--is 16777216 significant?

Reply:
Ryan, This test is named "nist_test_case_1", so I think these values come from some official documentation.

Itamar, consider putting a link to the definition for this test, or have a comment with an exact term that the reader could google to find it, or copy/paste the definition of the test from whatever doc it's coming from. The idea is: if this test fails in some far-flung future. Some poor engineer is going to need to figure out exactly what it was trying to do, so you need to give them the best possible clues